{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "# Aldy 4 Experimental Notebook\n",
    "---\n",
    "\n",
    "## Evaluated tools\n",
    "\n",
    "| Tool | Version |\n",
    "|------|-----|\n",
    "| Aldy | [v4.3](https://github.com/0xTCG/aldy) |\n",
    "| Aldy | [v3.3](https://pypi.org/project/aldy/3.3/) |\n",
    "| Astrolabe | [0.8.7.2](https://www.childrensmercy.org/genomesoftwareportal/Software/Details/4/) |\n",
    "| Cyrius | [1.1.1](https://github.com/Illumina/Cyrius) |\n",
    "| Stargazer | [1.0.8](https://stargazer.gs.washington.edu/stargazerweb/index.html) |\n",
    "| StellarPGx | [1.2.6](https://github.com/SBIMB/StellarPGx) |\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "pycharm": {
     "name": "#%% md\n"
    },
    "tags": []
   },
   "source": [
    "## Preprocessing\n",
    "\n",
    "### Stargazer: Creation of GATK DepthofCoverage Format (GDF) file\n",
    "`REF_FASTA` is the reference fasta used in samples' read alignment files and `BED` is the BED file used in subsetting regions of interest in the reference fasta.\n",
    "```bash\n",
    "module load GATK/4.2.6\n",
    "gatk --java-options \"-Xms20G -Xmx20G -XX:ParallelGCThreads=2\" DepthOfCoverage \\\n",
    "  -R ${REF_FASTA} \\\n",
    "  -I ${SAMPLE}.bam \\\n",
    "  -L ${BED} \\\n",
    "  -O ${SAMPLE} \\\n",
    "  --output-format TABLE\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "pycharm": {
     "name": "#%% md\n"
    },
    "tags": []
   },
   "source": [
    "## Technologies used in benchmarking datasets\n",
    "\n",
    "| Short-read datasets | Long-read datasets |\n",
    "| :-----: | :-----: |\n",
    "| 10X Genomics | PacBio HiFi |\n",
    "| Illumina WGS |\n",
    "| PGRNseq v.3<br>(targeted sequencing) |"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    },
    "tags": []
   },
   "source": [
    "## Genes of interest --- Genes against which Aldy 4 and other tools were benchmarked\n",
    "\n",
    "<!-- | Genes for short-read data benchmarking | Genes for long-read data benchmarking |\n",
    "| ----- | ----- |\n",
    "| CYP1A1 | CYP1A2 |\n",
    "| CYP1A2 | CYP2B6 |\n",
    "| CYP2A13 | CYP2C8 |\n",
    "| CYP2A6 | CYP2C9 |\n",
    "| CYP2B6 | CYP2C19 |\n",
    "| CYP2C8 | CYP2D6 |\n",
    "| CYP2C9 | CYP3A4 |\n",
    "| CYP2C19 | CYP3A5 |\n",
    "| CYP2D6 | CYP4F2 |\n",
    "| CYP2F1 | DPYD |\n",
    "| CYP2J2 | NUDT15 |\n",
    "| CYP2S1 | SLCO1B1 |\n",
    "| CYP3A4 | TPMT |\n",
    "| CYP3A5 |\n",
    "| CYP3A7 |\n",
    "| CYP3A43 |\n",
    "| CYP4F2 |\n",
    "| DPYD |\n",
    "| SLCO1B1 |\n",
    "| TPMT |\n",
    "-->\n",
    "\n",
    "| Short-read data genes | Aldy 4 & Aldy 3 | Stargazer | StellarPGx | Astrolabe | Cyrius |\n",
    "| ----- | :-----: | :-----: | :-----: | :-----: | :-----: |\n",
    "| CYP2D6 | &#10003; | &#10003; | &#10003; | &#10003; | &#10003; |\n",
    "| CYP2A6 | &#10003; | &#10003; | &#10003; |  |  |\n",
    "| CYP1A1 | &#10003; | &#10003; |  |  |  |\n",
    "| CYP1A2 | &#10003; | &#10003; | &#10003; |  |  |\n",
    "| CYP2A13 | &#10003; | &#10003; |  |  |  |\n",
    "| CYP2B6 | &#10003; | &#10003; | &#10003; |  |  |\n",
    "| CYP2C8 | &#10003; | &#10003; |  | &#10003; |  |\n",
    "| CYP2C9 | &#10003; | &#10003; | &#10003; | &#10003; |  |\n",
    "| CYP2C19 | &#10003; | &#10003; | &#10003; | &#10003; |  |\n",
    "| CYP2J2 | &#10003; | &#10003; |  |  |  |\n",
    "| CYP2S1 | &#10003; | &#10003; |  |  |  |\n",
    "| CYP3A4 | &#10003; | &#10003; | &#10003; |  |  |\n",
    "| CYP3A5 | &#10003; | &#10003; | &#10003; |  |  |\n",
    "| CYP3A7 | &#10003; | &#10003; |  |  |  |\n",
    "| CYP3A43 | &#10003; | &#10003; |  |  |  |\n",
    "| CYP4F2 | &#10003; | &#10003; | &#10003; | &#10003; |  |\n",
    "| DPYD | &#10003; | &#10003; |  |  |  |\n",
    "| SLCO1B1 | &#10003; | &#10003; |  | &#10003; |  |\n",
    "| TPMT | &#10003; | &#10003; |  | &#10003; |  |\n",
    "\n",
    "| Long-read data genes | Aldy 4 | Astrolabe |\n",
    "| ----- | ----- | ----- |\n",
    "| CYP2D6 | &#10003; | &#10003; |\n",
    "| CYP1A2 | &#10003; | |\n",
    "| CYP2B6 | &#10003; | |\n",
    "| CYP2C8 | &#10003; | &#10003; |\n",
    "| CYP2C9 | &#10003; | &#10003; |\n",
    "| CYP2C19 | &#10003; | &#10003; |\n",
    "| CYP3A4 | &#10003; | |\n",
    "| CYP3A5 | &#10003; | |\n",
    "| CYP4F2 | &#10003; | &#10003; |\n",
    "| DPYD | &#10003; | |\n",
    "| NUDT15 | &#10003; | |\n",
    "| SLCO1B1 | &#10003; | &#10003; |\n",
    "| TPMT | &#10003; | &#10003; |"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "## Genotyping\n",
    "\n",
    "### Aldy 4\n",
    "\n",
    "#### 10X\n",
    "```bash\n",
    "  aldy genotype -p 10x -g ${gene} ${SAMPLE}.bam\n",
    "```\n",
    "\n",
    "#### Illumina WGS\n",
    "```bash\n",
    "  aldy genotype  -p wgs -g ${gene} ${SAMPLE}.bam\n",
    "```\n",
    "\n",
    "#### PGRNseq v.3\n",
    "```bash\n",
    "  aldy genotype -p pgx3 -g ${gene} ${SAMPLE}.bam\n",
    "```\n",
    "\n",
    "#### PacBio\n",
    "```bash\n",
    "  aldy genotype -p pacbio-hifi-targeted -g ${gene} ${SAMPLE}.bam\n",
    "```\n",
    "\n",
    "The profile `pacbio-hifi-targeted-twist` was used for the second dataset.\n",
    "\n",
    "\n",
    "### Aldy 3\n",
    "\n",
    "#### 10X\n",
    "```bash\n",
    "  aldy genotype -p 10x -g ${gene} ${SAMPLE}.bam\n",
    "```\n",
    "\n",
    "#### Illumina WGS\n",
    "```bash\n",
    "  aldy genotype -p wgs -g ${gene} ${SAMPLE}.bam\n",
    "```\n",
    "\n",
    "#### PGRNseq v.3\n",
    "```bash\n",
    "  aldy genotype -p pgx3 -g ${gene} ${SAMPLE}.bam\n",
    "```\n",
    "\n",
    "\n",
    "### Astrolabe\n",
    "In the following commands, `REF_FASTA` is the reference fasta used in samples' read alignment files and `BED` is the BED file used in subsetting regions of interest in the reference fasta. \n",
    "\n",
    "VCF files were generated as follows:\n",
    "\n",
    "```bash\n",
    "bcftools mpileup -f ${REF_FASTA} -Ou $1 | bcftools call -o ${SAMPLE}.raw.vcf -mv -Ob --threads 4\n",
    "bcftools view -i \"%QUAL>=20\" ${SAMPLE}.raw.vcf -o ${SAMPLE}.vcf.gz -O v\n",
    "```\n",
    "\n",
    "#### 10X / Illumina WGS / PGRNseq v.3\n",
    "```bash\n",
    "    ${ASTROLABE_PATH}/run-astrolabe.sh  -conf ${ASTROLABE_PATH}/astrolabe.ini \\\n",
    "                                    -inputVCF ${SAMPLE}.vcf \\\n",
    "                                    -inputBam ${SAMPLE}.bam \\\n",
    "                                    -skipVcfQC \\\n",
    "                                    -skipBamQC \\\n",
    "                                    -outFile ${SAMPLE}_${gene}.log \\\n",
    "                                    -intervals ${BED} \\\n",
    "                                    -fasta ${REF_FASTA} \\\n",
    "                                    -targets ${gene}\n",
    "```\n",
    "\n",
    "#### PacBio\n",
    "```bash\n",
    "    ${ASTROLABE_PATH}/run-astrolabe.sh  -conf ${ASTROLABE_PATH}/astrolabe.ini \\\n",
    "                                    -ref GRCh38 \\\n",
    "                                    -inputVCF ${SAMPLE}.vcf \\\n",
    "                                    -inputBam ${SAMPLE}.bam \\\n",
    "                                    -skipVcfQC \\\n",
    "                                    -skipBamQC \\\n",
    "                                    -outFile ${SAMPLE}_${gene}.log \\\n",
    "                                    -intervals ${BED} \\\n",
    "                                    -fasta ${REF_FASTA} \\\n",
    "                                    -targets ${gene}\n",
    "```\n",
    "\n",
    "\n",
    "### Cyrius\n",
    "\n",
    "#### 10X / Illumina WGS\n",
    "```bash\n",
    "    python3 Cyrius/star_caller.py --manifest ${LIST_OF_SAMPLE_BAMS} \\\n",
    "     --threads 16 \\\n",
    "     --genome 37 \\\n",
    "     --prefix result-${out_dir} --outDir outputs/${out_dir}\n",
    "```\n",
    "\n",
    "### Stargazer\n",
    "Used software packages: \n",
    "| Dependency | Version |\n",
    "|-----|-----|\n",
    "| Python | 3.6 |\n",
    "| R | 3.5 |\n",
    "| Beagle | 5.4 (instead of the default v5.1) |\n",
    "#### 10X / Illumina WGS\n",
    "```bash\n",
    "# Data type: wgs; control type: custom, with CYP2D8 gene as the control region (same as Aldy)\n",
    "    stargazer.py genotype -d wgs \\\n",
    "    -o ${SAMPLE}_${gene} \\\n",
    "    -t ${gene} \\\n",
    "    --vcf ${SAMPLE}.vcf --gdf ${SAMPLE}.gdf \\\n",
    "    --control_type custom --region 22:42547463-42548249\n",
    "```\n",
    "\n",
    "#### PGRNseq v.3\n",
    "We could get Stargazer to run in `-d ts` mode and we tried running many combinations before we decided on running it using WGS mode. The explanation is given in the next section.\n",
    "\n",
    "```bash\n",
    "# Data type: wgs; control type: custom, with CYP2D8 gene as the control region (same as Aldy)\n",
    "    stargazer.py genotype -d wgs \\\n",
    "    -o ${SAMPLE}_${gene} \\\n",
    "    -t ${gene} \\\n",
    "    --vcf ${SAMPLE}.vcf \\\n",
    "    --control_type custom --region 22:42547463-42548249\n",
    "```\n",
    "\n",
    "### StellarPGx\n",
    "\n",
    "#### 10X / Illumina WGS\n",
    "    nextflow run main.nf -profile standard --format compressed --build b37 --gene ${gene} --in_bam  ${SAMPLE}*{cram,crai}\\\"\"+\" --out_dir ${OUT_DIR}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "#  Running Stargazer on targeted sequencing data in `-d ts` mode\n",
    "\n",
    "Stargazer performs pharmacogene genotyping of a sample by taking the VCF (Variant Call Format) and the GDF (GATK's `DepthOfCoverage` Format) files of the sample as input. (note that Stargazer provides a script that can extract the VCF/GDF files from a sample BAM files).\n",
    "\n",
    "We tried running Stargazer on the PGRNseq v.3 samples using the recommended data type argument (`-d ts`) in multiple modes. However, we were unable to observe satisfactory results. The following command:\n",
    "\n",
    "    stargazer.py genotype -o OUTPUT_PREFIX -d ts -t TARGET_GENE --vcf VCF --gdf GDF --control_type cnsr\n",
    "\n",
    "failed with the following error message: \n",
    "\n",
    "    Genotyping with TS data requires at least five samples.\n",
    "\n",
    "We tried concatenating all VCF and GDF files (in the same sample order in the joint VCF and joint GDF files), and running Stargazer on the joint VCF + GDF combination.\n",
    "The default `cnsr` (copy-number stable region) as the `control_type` was not available for all genes. Instead, we tried using `CYP2D8` as the custom control region:\n",
    "\n",
    "    stargazer.py genotype -o OUTPUT_PREFIX -d ts -t TARGET_GENE --vcf JOINT_VCF --gdf JOINT_GDF --control_type custom --region 22:42547463-42548249\n",
    "\n",
    "However, this failed with the following error message: \n",
    "\n",
    "    TypeError: Something bad happened during SV detection!\n",
    "\n",
    "We also tried running Stargazer through the VCF-only mode which detects SNVs but not structural variants. Almost every sample, for almost every gene of interest, had a genotype call of `*1/*1`. The only genes that seemed to work in the VCF-only mode were `CYP2D6` and `SLCO1B1`; however, structural variant discovery could not be done without the GDF input. The run command was:\n",
    "\n",
    "    stargazer.py genotype -o OUTPUT_PREFIX -d ts -t TARGET_GENE --vcf JOINT_VCF --control_type custom --region 22:42547463-42548249\n",
    "\n",
    "Finally, we were able to get the satisfactory results when we ran Stargazer in WGS mode (`-d wgs`), turning off the structural variant detection, on the PGRNseq dataset using the following command:\n",
    "\n",
    "    stargazer.py genotype -o OUTPUT_PREFIX -d wgs -t TARGET_GENE --vcf VCF --control_type custom --region 22:42547463-42548249\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "## Parsing the results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    },
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "'''\n",
    "get result of astrolabe on samples\n",
    "'''\n",
    "def get_astrolabe_result(gene, bam_name,tech):\n",
    "    gene = gene.upper()\n",
    "    f_name = 'GR_outputs/{}/astrolabe_{}_{}.txt'.format(tech,bam_name,gene)\n",
    "\n",
    "    fp = open(f_name,'r')\n",
    "    last = fp.read().splitlines()[-1].split('\t')[1]\n",
    "    last = last.replace(gene,'')\n",
    "    last = last.replace('*','')\n",
    "    last = last.replace(' or ',';')\n",
    "    res = last\n",
    "    times = re.findall('\\dx\\d',res)\n",
    "    for s in times:\n",
    "        print(s)\n",
    "        first = s[0]\n",
    "        second = s[-1]\n",
    "        newstr = first\n",
    "        for i in range(int(second)-1):\n",
    "            newstr=newstr+'+'+first\n",
    "        res = res.replace(s,newstr)\n",
    "\n",
    "    res = res.replace(',', ';')\n",
    "    res = res.replace('./.','w/ ')\n",
    "    return res\n",
    "\n",
    "\n",
    "'''\n",
    "get the result of stargazer\n",
    "'''\n",
    "def get_stargazer_result(gene, bam_name, tech):\n",
    "    gene = gene.lower()\n",
    "    f_name = 'GR_outputs/{}/stargazer_{}_{}.txt'.format(tech,bam_name,gene)\n",
    "\n",
    "    fp = open(f_name, 'r')\n",
    "    last = fp.read().splitlines()[-1].split('\t')\n",
    "    f,s = last[2],last[3]\n",
    "    f = f.replace('*','')\n",
    "    s = s.replace('*','')\n",
    "    res = '{}/{}'.format(f,s)\n",
    "    times = re.findall('\\dx\\d',res)\n",
    "    for s in times:\n",
    "        print(s)\n",
    "        first = s[0]\n",
    "        second = s[-1]\n",
    "        newstr = first\n",
    "        for i in range(int(second)-1):\n",
    "            newstr=newstr+'+'+first\n",
    "        res = res.replace(s,newstr)\n",
    "    if res.find('./.')!=-1:\n",
    "        cand1 = last[4]\n",
    "        cand2 = last[5]\n",
    "        if len(cand1.split(','))>1:\n",
    "            cand1 = '[{}]'.format(cand1)\n",
    "        if len(cand2.split(','))>1:\n",
    "            cand2 = '[{}]'.format(cand2)\n",
    "        res = '{}/{}'.format(cand1,cand2)\n",
    "        res = res.replace('*','')\n",
    "    return res\n",
    "\n",
    "'''\n",
    "get result of cyrius on samples\n",
    "'''\n",
    "def get_cyrius_result(tech, bam_name):\n",
    "    file = open('GR_outputs/{}/cyrius.txt'.format(tech),'r')\n",
    "    for line in file.readlines():\n",
    "        if line.split('\\t')[0] == bam_name:\n",
    "            res = line.split('\\t')[1].strip().replace('*','')\n",
    "            return  res\n",
    "    return 'Not found'\n",
    "\n",
    "'''\n",
    "get the result of stellarpgx\n",
    "'''\n",
    "def get_stellarpgx_result(gene, tech, bam_name):\n",
    "    f_name = 'GR_outputs/{}/stellarpgx_{}_{}.txt'.format(tech,bam_name,gene)\n",
    "    with open(f_name, 'r') as handle:\n",
    "        copy_number = 0\n",
    "        allele = \"\"\n",
    "        for ln in handle:\n",
    "            if 'No_call' in ln:\n",
    "                return 0, 'No_call/No_call'\n",
    "            if (ln.startswith(\"Initially computed CN\")):\n",
    "                cn = int(ln.split(\"=\")[1].strip())\n",
    "            if (ln.startswith(\"*\")):\n",
    "                allele = ln.strip()\n",
    "                allele = allele.replace('*', '')\n",
    "            if  (ln.startswith(\"Likely background alleles\")):\n",
    "                allele = next(handle).strip().replace(\"[\",\"\").replace(\"]\",\"\")\n",
    "                allele = allele.replace('*','')\n",
    "            if 'or' in allele:\n",
    "                k = allele.split('or')\n",
    "                for s in k:\n",
    "                    s = s.replace(\" \",\"\")\n",
    "                allele = ';'.join(k)\n",
    "    return copy_number, allele\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "toc": {
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
